### ISSUE TITLE
why are `--device`s bind-mounted?

### ISSUE THREAD

**Issue by M1cha** [containers/podman#16827]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

Right now there's no way for a `--userns=auto` container to use a UART-device (`/dev/ttyUSB*`) due to ownership and permissions of the host file. IMO this could be easily solved by copying the file into the containers `/dev` filesystem and giving it proper ownership. That would also allow using it in multiple containers without having to make it world-readwrite-able on the host.

**Comment by M1cha** [comment]
@flouthoc I'm running podman as root. You're right in that the documentation says that this is done for rootless: `In rootless mode, the new device is bind mounted in the container from the host rather than Podman creating it within the container space.`.

So is it actually a bug that this is done for rootful with userns=auto as well?

About rootless: I thought about using idmap for this but apparently devtmpfs doesn't set `FS_ALLOW_IDMAP`.

**Comment by M1cha** [comment]
here's also a quick way to reproduce this in case you wanna see for yourself:(you can use any dev, I used /dev/mem because it's more likely that you have that than ttyUSB):
`# podman run --userns=auto --rm -it --device /dev/mem:/dev/mem:rwm alpine ls -lah /dev/mem`

and it shows `crw-r-----    1 nobody   nobody      1,   1 Dec 10 08:40 /dev/mem` instead of being owned by root.

**Comment by giuseppe** [comment]
yes, a rootless user cannot create device inside the container since mknod is blocked.  So they are bind mounted from the host

**Comment by mheon** [comment]
I would think that we could do regular device nodes in `--userns=auto`, though, which is what is in use? This still seems like a bug to me.

**Comment by giuseppe** [comment]
sorry I clicked too fast.  I had another look and I don't see what we can do from Podman.

Podman is doing the right thing:

```
    "devices": [
      {
        "path": "/dev/mem",
        "type": "c",
        "major": 1,
        "minor": 1,
        "fileMode": 8608,
        "uid": 200000,
        "gid": 200009
      }
    ],
```

but the OCI runtime cannot create a device while being in a user namespace.  I've tried with both crun and runc and they have the same behavior.

The user namespace must be created earlier because the mount namespace must be owned by the correct user namespace, and we need the mounts created before we create the devices (e.g. /dev is a tmpfs and must be created before we create devices).

We could probably do some really ugly things like forking a different process that joins only the mount namespace but not the user namespace to create these devices, but it seems like a not necessary complication.

In any case, the issue cannot be addressed in podman.

**Comment by M1cha** [comment]
the runtime gets run in the root namespace though right? If yes this should be fixable within e.g. crun.

**Comment by rhatdan** [comment]
Rootless user can not create a device whether or not it is in a user namespace.
```
$ podman unshare
# mknod /tmp/dan c 0 0
# rm /tmp/dan
```
You are only allowed to create a device with major/minor number 0 0.

If you try to create /dev/null, it fails
```
# ls -l /dev/null 
crw-rw-rw-. 1 nobody nobody 1, 3 Dec 13 16:39 /dev/null
# mknod /tmp/dan c 1 3
mknod: /tmp/dan: Operation not permitted
```

**Comment by rhatdan** [comment]
Ok, I just re-read the description. This is crun running within a user namespace, feeling to be able to create device nodes.
This really stinks that this is not possible.

**Comment by giuseppe** [comment]
it is not possible, as I've explained in https://github.com/containers/podman/issues/16827#issuecomment-1351717523

when we get to create devices we are already in new user&mount namespace, so we lost the ability of doing mknod.

The only way of achieving it is through a helper program that runs on the host and joins only the mount namespace, but it seems like a overcomplicated design for such a minor issue.

**Comment by M1cha** [comment]
@giuseppe your statement doesn't quite reflect the code I see within crun: https://github.com/containers/crun/blob/313fc8a6ad3a8c758d3fcfc62898bf199f33f0a3/src/libcrun/container.c#L2313

They start the container as a separate process and go through several init-phases by waiting on sync points both within the container process and the host process. Using one of those sync point in the host process to do something like `fork(); setns(), mknod()` sounds trivial to me.

Also, I disagree that this is a minor issue since it makes `--device` completely unusable for rootful `--userns=auto`. You can't know which used to chown the device to since the user mappings are managed by podman and different for ever start. Also, you don't want to give access to `others` since that'd be a huge security risk.

**Comment by giuseppe** [comment]
> They start the container as a separate process and go through several init-phases by waiting on sync points both within the container process and the host process. Using one of those sync points in the host process to do something like `fork(); setns(), mknod()` sounds trivial to me.
> 
> Also, I disagree that this is a minor issue since it makes `--device` completely unusable for rootful `--userns=auto`. You can't know which uid/gid to chown the device to since the user mappings are managed by podman and different for every start. Also, you don't want to give access to `others` since that'd be a huge security risk.

I wasn't aware it was such a major problem.

We can move the discussion to the crun repository since the fix must be implemented there.

Would you mind opening a PR?

**Comment by rhatdan** [comment]
We can transfer this issue over there.
